File: Problems/sockMerchant.js
/****************************************************************/
/*Problem: Sock Merchant (HR)  ********/
/****************************************************************/
/*
-- Summary:
John works at a clothing store. He has a large pile of socks that he must pair by color for sale. Given an array of integers representing the color of each sock, determine how many pairs of socks with matching colors there are.
For example, there are n=7 socks with colors ar=[1,2,1,2,1,3,2]. There is one pair of color 1 and one of color 2. There are three odd socks left, one of each color. The number of pairs is 2.
-- Input(s):
The first line contains an integer n, the number of socks represented in ar. The second line contains n space-separated integers describing the colors ar[i] of the socks in the pile.
-- Expected output(s):
Complete the sockMerchant function in the editor below. It must return an integer representing the number of matching pairs of socks that are available.
-- Constraints:
*/
function sockMerchant(n, ar) {
  let map = {};
  let pairs = 0;

  for (let i = 0; i < ar.length; i++) {
    const item = ar[i];
    if (!map[item]) {
      map[item] = 1;
    } else {
      map[item] += 1;
    }
  }
  //console.log(map);

  for (let i = 0; i < ar.length; i++) {
    const item = ar[i];
    if (map[item] % 2 === 0 && map[item] > 0) {
      pairs += 1;
      map[item] = map[item]-2;
    } else if (map[item] > 2) {
      pairs += 1;
      map[item] = map[item]-2;
    }
  }

  console.log(pairs);
}
sockMerchant(7,[1,2,1,2,1,3,2]);
sockMerchant(9,[10,20,20,10,10,30,50,10,20]);
sockMerchant(1,[10]);
sockMerchant(3,[10,10,10]);
sockMerchant(9,[1,20,25,10,11,30,50,12,1]);

File: Problems/Hackerrank/sockMerchant.cpp
/*********************************************************************************************/
/* Problem: Sock Merchant (HR) ********/
/*********************************************************************************************/
/*
--Problem statement:
  There is a large pile of socks that must be paired by color. Given an array of integers
  representing the color of each sock, determine how many pairs of socks with matching colors
  there are.
  > Example:
    n = 7         ar = [1,2,1,2,1,3,2]
    There is one pair of color 1 and one of color 2. There are three odd socks left, one of each
    color. The number of pairs is 2.

  > Input:
      int n: the number of socks in the pile
      int ar[n]: the colors of each sock

  > Output:
      int: the number of pairs

  > Constraints:
    1 <= n <= 100
    1 <= ar[i] <= 100 where 0 <= i < n

--Reasoning:
    Iterate over the input array mapping each element and checking if an element equal to the current
    element has already been mapped (i.e. socks[element]=true). If so, it means that we now have
    formed a new pair (the one mapped + the current), so add one more pair to the result and "remove"
    this one from the hash table. In case it has been mapped yet, add it to the hash table.

--Time complexity: O(n), where n is the size of the input array.

--Space complexity: O(n), due to the hash table created which, in the worst case, will have only
socks without pairs.
*/

#include <iostream>
#include <vector>
#include <unordered_map>

int sockMerchant(const std::vector<int> &ar)
{
  int pairs{0};
  std::unordered_map<int, bool> socks;

  for (const auto &a : ar)
  {
    if (!socks[a])
    {
      socks[a] = true;
    }
    else
    {
      ++pairs;
      socks[a] = false;
    }
  }

  return pairs;
}

int main()
{
  std::vector<int> all_socks{1, 2, 1, 2, 1, 3, 2};
  int pairs = sockMerchant(all_socks);
  std::cout << "Number of pairs - Expected: 2\t\tResult: " << pairs << (pairs == 2 ? "\tPASSED\n" : "\tFAILED\n");

  all_socks = {10, 20, 20, 10, 10, 30, 50, 10, 20};
  pairs = sockMerchant(all_socks);
  std::cout << "Number of pairs - Expected: 3\t\tResult: " << pairs << (pairs == 3 ? "\tPASSED\n" : "\tFAILED\n");

  all_socks = {4, 5, 5, 5, 6, 6, 4, 1, 4, 4, 3, 6, 6, 3, 6, 1, 4, 5, 5, 5};
  pairs = sockMerchant(all_socks);
  std::cout << "Number of pairs - Expected: 9\t\tResult: " << pairs << (pairs == 9 ? "\tPASSED\n" : "\tFAILED\n");

  return 0;
}